README.md
<p align="center">
  <b> psilib </b> &bull;
  <a href="https://github.com/sn3ksoftware/psicli">psicli</a> &bull;
    <a href="https://github.com/sn3ksoftware/psidex">psidex</a>
</p>

# psilib: The json-based backend for the psiman package manager
(formerly [spkg for Libterm](https://github.com/sn3ksoftware/sandpkg/tree/testing))

<img src="https://raw.githubusercontent.com/sn3ksoftware/psilib/master/psilib_logo.png" alt="psiman_logo:Ψ" width="200"/>

Welcome to the repo for psilib, a json-based backend for the [psiman](https://github.com/sn3ksoftware/psiman) package manager.

It uses the [psidex](https://github.com/sn3ksoftware/psidex) system to distribute packages.

Platforms supported:

| Platform  | Version of psilib |
| --- | --- |
| [Libterm](https://github.com/ColdGrub1384/LibTerm) | ![GitHub release (latest SemVer including pre-releases)](https://img.shields.io/github/v/release/sn3ksoftware/psilib?include_prereleases&sort=semver) |
| [Pythonista](http://omz-software.com/pythonista/) | ![GitHub release (latest SemVer including pre-releases)](https://img.shields.io/github/v/release/sn3ksoftware/psilib?include_prereleases&sort=semver) |
| Linux (Generic) | -WIP- |
| Windows | -WIP- |

For more info, take a look at the [documentation](https://github.com/sn3ksoftware/psilib/wiki).

# TODO
The list below is as follows:

**(M)** denotes a major change to psilib/psiman code.

**(m)** denotes a minor change to psilib/psiman code.

**(p)** denotes a patch for a bug within psilib/psiman code.

- [x] (M) Add semantic versioning to psilib.package.install()
- [ ] (M) Add code to handle dependencies
- [x] (M) Implement the whole Pythonista Script Index specification, i.e support multiple 
packages using the same json file
- [ ] (M) Get psilib to a stable API release (0.1.0)
- [ ] (M) Add global cache to store indexes of muitiple repos (i.e, no need to switch repositories.)
- [x] (m) Make setrepo() and delrepo() non-interactive 
- [x] (m) Update meta-version in all packages to reflect API changes
- [x] (p) Make the cli-wrapper for psilib(psiman) less buggy
- [x] (p) Format output of the cli-wrapper properly

# P.S
The code from [spkg for Libterm](https://github.com/sn3ksoftware/sandpkg/tree/testing) was _heavily_ modified to use the json-based `psidex` system.
Its original source code, including the psiman package manager, is under the MIT License.


CHANGELOG.md
# Changelog

### psilib v3.3.0-alpha:
- Indexes are now downloaded for all repos at once. This means that there is no need to switch repos to get certain packages. (Also, the upgrade function now works.)
- This incidentally also removes the need for psilib.repo.set().
- psilib.repo.fetch() has been moved to psilib.repo.update().
- psilib.repo.update() now returns True if all indexes were downloaded sucessfully.
- Package installation was broken due to different possible configurations for the Script Definition file. It now uses ywangd’s original system.

### psilib v3.2.0-alpha:
- Alpha-stage support for ywangd's older implementation of the Pythonista Script
Index. Because of the inconsistency of package formats, this is VERY unstable
and NOT gurenteed to work or install command-line scripts. More support will be added
to a new bugfix later on (problably v3.2.1-alpha.)
- Added support for packages hosted within a Github Gist. Does not work on muiti-file
gists.
- Packages can now be one of these formats: a singular .py or .pyui file, or
a zipfile (that can contain dev-specific dependencies).
The `type` key in the script.json file must be set to either 'module' or 'script'.
(Scripts will be extracted/moved to `psilib.config.script_store`, and modules
extracted to `psilib.config.site_packages`.
- Using `psiman setrepo` will now automatically download the index from the new repo.

### psilib v3.1.0-alpha:
- Lots of bugfixes for both psilib and
the psiman cli-wrapper.
- psilib now supports semantic versioning as package versions.
- Package versions are now compared to a local file (versions.json) so that packages will NOT be redownloaded if the package in the repo is at the same version as the local counterpart.
- Packages will be installed to site-packages if it is a module.
- New function, psilib.package.upgrade(), which downloads and installs all packages with a newer version available in the repository.
- New function, psilib.package.search(), that finds/lists packages based on name, type(module/script), and branch(stable/rolling).
- psiman.repo.set() and psiman.repo.rem() are now non-interactive.
Use psiman.repo.list() to list repos in repolist and then pass the 'N'th
repository number to set() and rem().

### psilib v3.0.0-alpha:
THE MODULAR UPDATE
The singular psiman script has now been split
into two: A backend library (psilib) that
handles low-level functions and a frontend
script (psiman) that interfaces with psilib.
The split was done as:
- The code is now easier to debug as it is
not a continueous wall of text.
- It allows package installs, etc. to be scripted.
- psilib can have seperate dev cycles from
psiman, and vice versa.
- New functions can be updated easily.


psiman.py
#!/usr/bin/env python3

"""
0000  00000 00000 0   0   0   00  0
00  0 00      0   00 00  0 0  00  0
00000 00000   0   0 0 0 00000 0 0 0
00       00   0   0   0 0   0 0  00
00    00000 00000 0   0 0   0 0  00
(psiman)
The cli-wrapper for psilib, allowing use
from the command-line.

Copyright (c) 2019 Ong Yong Xin
Open-sourced under the MIT License.
To print out the license, execute the below:
import psilib as psi; psi.misc.license()

Source:
https://github.com/sn3ksoftware/psiman
"""

# Import functions from stdlib
import sys
from inspect import getargspec

# Import functions from psilib
from psilib.repo import add, set, rem, fetch
from psilib.package import install, remove, list, desc
from psilib.misc import credits, license
from psilib.utils import ansi

# If there are no arguements, provide user help
def getParameters(args):
    if len(args) == 1:
        print(
        "\nusage: psiman (action) [options] <input>\n"
        "Type [psiman -h] for help.\n"
        )
        return None
    else:
        arg1_raw = (sys.argv[1])
        return(arg1_raw)

# Get first arguement from params
arg1 = getParameters(sys.argv)
argall = sys.argv

# Declare wrapper-specific variables.
invalid_cmd = ansi(31) + '\nE: Invalid/unknown command. Try again.\n\n usage: psiman (action) [options] <input>\nType [psiman -h] for help.\n' + ansi(0)

# Dictionary used to map functions.
func_dict = {
    "-h": "help",
    "-c": "credits",
    "-ar": "add",
    "-sr": "set",
    "-dr": "rem",
    "-i": "install",
    "-r": "remove",
    "-l": "list",
    "-d": "desc",
    "-f": "fetch",
}

# Help function. This is specific to the wrapper itself.
def help():
    print(
    """
Printing help...
\n-PACKAGE MANAGEMENT-\n
\t[-i], {package}: Install new packages.
\t[-r], {package}: Remove specified package.
\t[-l], <repo/lcl>: List packages in repo/installed packages.
\t[-d], {package}: Print description of package specified(if any).
\n-REPOSITORY MANAGEMENT-\n
\t[-ar], {repo}: Adds new repository.
\t[-sr]: Sets main repository to use.
\t[-dr]: Removes repository from local list.
\t[-f]: Fetch latest index file from main repository.
\n-MISCELLANEOUS-\n
\t[-init], <repo/pkg>: Restores repo to defaults/deletes\n\tall spkg-installed packages(wip).
\t[-h]: Prints help screen.
\t[-c]: Print version and credits.
    """
    )

# Special function that pretty-prints input based on type.
def prettyprint(text):
    input_type = type(text)
    # Define supported types to print
    isString = (input_type is str)
    isNumber = (input_type is int or input_type is float)
    isIteratable = (input_type is list or input_type is tuple)
    isBool = (input_type is bool)
    isDict = (input_type is dict)
    if isString:
        print(ansi(32), text, ansi(0))
    elif isNumber:
        print(ansi(32), str(text), ansi(0))
    elif isBool:
        if text == True:
            print(ansi(32, bold=True), "Sucess!", ansi(0))
        elif text == False:
            print(ansi(31, bold=True), "Failure!", ansi(0))
    elif isDict:
        print("\n")
        for key, val in text.items():
            # Unpack if there is a nested dict
            if type(val) is dict:
                print(ansi(34, bold=True) + key + ":")
                for keyn, valn in val.items():
                    print(ansi(32, bold=True), keyn, ": ", ansi(0), ansi(36), valn, ansi(0), sep="")
            else:
                print(ansi(32, bold=True), key, ": ", ansi(0), ansi(36), val, ansi(0), sep="")
        print("\n")
    else:
        # A list/tuple.
        try:
           print(ansi(32))
           for i in text:
               print(i)
           print(ansi(0))
        except TypeError:
            # None type?
            print(text)

# Define main function.
def main():
    # Try to get function name. If it fails, print a error message.
    try:
        cmd_name = func_dict[arg1]
    except KeyError:
        if len(sys.argv) == 1:
            pass
        else:
            print(invalid_cmd)
    else:
        # Check if function requires args.
        # By default, sys.argv[2] will be
        # passed to the function.
        func = eval(cmd_name)
        try:
            if func() == None:
                pass
            else:
                prettyprint(func())
        except TypeError:
            try:
                prettyprint(func(sys.argv[2]))
            except IndexError:
                print(ansi(31))
                print(
                f"E: action {cmd_name} requires input, i.e\n"
                f"psiman {arg1} <input>\n"
                "Type (psiman -h) for more info."
                )
                print(ansi(0))

# Standard boilerplate to call the main() function to begin
# the program.
if __name__ == '__main__':
    main()

install.py
#!/usr/bin/env python3

"""
0000  00000 00000 00000 0     00000
00  0 00      0   0     0       0  
00000 00000   0   0     0       0
00       00   0   0     0       0
00    00000 00000 00000 00000 00000
(psicli)
- INSTALLER EDITION -

v2.1.4

A blob that installs psicli (the cli-wrapper
for psilib), and psilib.

Copyright (c) 2019 Ong Yong Xin
Open-sourced under the MIT License.

Source:
https://github.com/sn3ksoftware/psicli
"""

# Import needed modules.
import os
import sys
import shutil
import zipfile
import platform
import requests

# Get options from sys.argv


def chkarg1():
	try:
		arg1 = sys.argv[1]
	except IndexError:
		return False
	else:
		return True
		
# Check install platform


def chkplat():
	if platform.machine().startswith("iP"):
		# iOS
		try:
			import objc_util
		except ImportError:
			# Running on Libterm, return path to scripts and psicfg.
			inslist = ["~/Library/bin",
			"~/Documents/.psicfg",
			"Libterm"]
			return inslist
		else:
			# Running on Pythonista, check if StaSh is installed.
			try:
				import stash
			except ImportError:
				print("\nE: StaSh is not installed. Please install StaSh first"
				"with [import requests as r; exec(r.get('https://bit.ly/"
				"get-stash').text)].")
				exit()
			else:
				print("\nInstaller is running on Pythonista, "
				"StaSh is installed.")
				inslist = ["~/Documents/bin",
				"~/Documents/.psicfg",
				"Pythonista"]
				return inslist
	else:
		# Not running on iOS: Platform not supported.
		raise NotImplementedError("\nE: Your platform is not supported.")
		
# Declare needed vars.
paths = chkplat()
home = os.path.expanduser("~/Documents/")
script_p = os.path.expanduser(paths[0])
psicfg_p = os.path.expanduser(paths[1])
tmp = psicfg_p + "/tmp/"
site_packages = next(p for p in sys.path if 'site-packages' in p)
plat = paths[2]
	
# Define extractfile function
# Copied from zip2 command


def extractfile(filepath, dest):
	# Parse filepath
	filepath_p = os.path.expanduser(filepath)
	dest_p = os.path.expanduser(dest)
	print("\nExtracting...")
	with zipfile.ZipFile(filepath_p, 'r') as zip:
		try:
			zip.extractall(path=dest_p)
		except zipfile.BadZipfile:
			print("\nE: Zip file is bad (corrupted?)\nAbort.")
		else:
			print("\nSucessfully extracted.")
			
# Define download function.


def dl(url, path=""):
	# Parse out path to be usable
	path_p = os.path.expanduser(path)
	# Split url and get filename
	filename = url.split("/")[-1]
	file = requests.get(url).content
	with open(path_p + filename, "wb") as f:
		f.write(file)
		
# Download psicli from repo and unzip to script folder.


def getpsi():
	resource_path = home + "resources.zip"
	print("\nStarting install of PSI...")
	
	# Download resources file and extract to
	# .psicfg
	dl("https://raw.githubusercontent.com/sn3ksoftware/psilib/master/resources.zip", path=home)
	try:
		os.mkdir(home + ".psicfg")
	except OSError:
		shutil.rmtree(home + ".psicfg")
		os.mkdir(home + ".psicfg")
	
	extractfile(resource_path, home + ".psicfg")
	
	print("\nGetting psicli/psilib...")
	
	# Download psicli and psilib
	dl("https://raw.githubusercontent.com/sn3ksoftware/psidex/master/packages/rolling/psicli.zip", path=tmp)
	dl("https://raw.githubusercontent.com/sn3ksoftware/psidex/master/packages/rolling/psilib.zip", path=tmp)
	
	try:
		os.mkdir(site_packages + "/psilib")
	except OSError:
		print("\nW: psilib already exists.\nReplacing...")
		shutil.rmtree(site_packages + "/psilib")
		os.mkdir(site_packages + "/psilib")
	else:
		pass
	
	# Extract psilib to site-packages, and
	# move psicli.py to script path.
	# If user specifes [-m] option, install
	# psilib module only.
	if chkarg1():
		if sys.argv[1] == "-m":
			extractfile(tmp + "psilib.zip", site_packages + "/psilib/")
			print("\nInstalled psilib only.")
	else:
		extractfile(tmp + "psilib.zip", site_packages + "/psilib/")
		extractfile(tmp + "psicli.zip", script_p)
		# Clean up tmp folder and delete
		# resources.zip
		for file in os.listdir(tmp):
			tmpfile = os.path.join(tmp, file)
			os.remove(tmpfile)
		os.remove(resource_path)
		print("\nInstalled both psicli and psilib.")
			
if __name__ == '__main__':
	getpsi()
